File: frontend/src/contracts/apiKeysContract.ts
/**
 * API-Keys-Contract v1 — Zod-Spiegel.
 *
 * Hand-gepflegt, 1:1 zu schemas/api-key*.schema.json. Änderungen am
 * Pydantic-Modell (backend/app/contracts/api_keys_contract.py) →
 * Schema-Dump → diese Datei synchronisieren.
 */
import { z } from "zod";

// === ApiKeyScope ===
export const ApiKeyScopeSchema = z.enum(["read", "write", "admin"]);
export type ApiKeyScope = z.infer<typeof ApiKeyScopeSchema>;

// === ApiKeyStatus ===
export const ApiKeyStatusSchema = z.enum(["active", "revoked"]);
export type ApiKeyStatus = z.infer<typeof ApiKeyStatusSchema>;

// === ApiKeyModel (persistierte Repräsentation, ohne Klartext) ===
export const ApiKeyModelSchema = z
  .object({
    id: z.string().min(1),
    label: z.string().min(1).max(120),
    prefix: z
      .string()
      .min(12)
      .max(12)
      .regex(/^ago_[0-9a-f]{8}$/),
    scopes: z.array(ApiKeyScopeSchema).min(1),
    status: ApiKeyStatusSchema,
    hashed_token: z.string(),
    created_at: z.string(),
    last_used_at: z.string().nullable().optional(),
    revoked_at: z.string().nullable().optional(),
  })
  .strict();
export type ApiKeyModel = z.infer<typeof ApiKeyModelSchema>;

// === ApiKeyCreateRequest ===
export const ApiKeyCreateRequestSchema = z
  .object({
    label: z.string().min(1).max(120),
    scopes: z.array(ApiKeyScopeSchema).min(1),
  })
  .strict();
export type ApiKeyCreateRequest = z.infer<typeof ApiKeyCreateRequestSchema>;

// === ApiKeyCreateResponse (Klartext-Token erscheint hier genau einmal) ===
export const ApiKeyCreateResponseSchema = z
  .object({
    key: ApiKeyModelSchema,
    token: z.string().regex(/^ago_[0-9a-f]{48}$/),
  })
  .strict();
export type ApiKeyCreateResponse = z.infer<typeof ApiKeyCreateResponseSchema>;

// === ApiKeysListResponse ===
export const ApiKeysListResponseSchema = z
  .object({
    items: z.array(ApiKeyModelSchema),
    total: z.number().int().min(0),
  })
  .strict();
export type ApiKeysListResponse = z.infer<typeof ApiKeysListResponseSchema>;

File: backend/app/contracts/api_keys_contract.py
"""API-Keys-Contract v1 (Pydantic v2).

Single-Workspace-Scope: API-Schlüssel sind keine User-Identity, sondern
Workspace-Zugänge mit Scopes (read|write|admin). Klartext-Token (`ago_<48-hex>`)
wird nur einmal im CreateResponse zurückgegeben; persistiert wird nur der Prefix
zur UI-Anzeige.

Status:
  active   — Schlüssel ist verwendbar.
  revoked  — Manuell zurückgezogen, kein erneutes Aktivieren.

Siehe ADR-Slice G2 + docu/2026-05-14-design-v4-slice-g2-api-keys-worklog.md.
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

ApiKeyScope = Literal["read", "write", "admin"]
ApiKeyStatus = Literal["active", "revoked"]

_STRICT = ConfigDict(extra="forbid")

# Klartext-Token-Format: ago_<48-hex>
TOKEN_PREFIX_LENGTH = 12  # "ago_" + 8 Hex-Zeichen


class ApiKeyModel(BaseModel):
    """Persistierte Repräsentation eines API-Schlüssels (ohne Klartext)."""

    model_config = _STRICT

    id: str = Field(min_length=1)
    label: str = Field(min_length=1, max_length=120)
    prefix: str = Field(
        min_length=TOKEN_PREFIX_LENGTH,
        max_length=TOKEN_PREFIX_LENGTH,
        pattern=r"^ago_[0-9a-f]{8}$",
    )
    scopes: list[ApiKeyScope] = Field(min_length=1)
    status: ApiKeyStatus
    created_at: datetime
    last_used_at: Optional[datetime] = None
    revoked_at: Optional[datetime] = None


class ApiKeyCreateRequest(BaseModel):
    """Eingabe für POST /api/api-keys."""

    model_config = _STRICT

    label: str = Field(min_length=1, max_length=120)
    scopes: list[ApiKeyScope] = Field(min_length=1)


class ApiKeyCreateResponse(BaseModel):
    """Antwort auf POST /api/api-keys — Klartext-Token erscheint hier genau einmal."""

    model_config = _STRICT

    key: ApiKeyModel
    token: str = Field(pattern=r"^ago_[0-9a-f]{48}$")


class ApiKeysListResponse(BaseModel):
    """Antwort auf GET /api/api-keys."""

    model_config = _STRICT

    items: list[ApiKeyModel]
    total: int = Field(ge=0)
